Add time-optimal trajectory parameterization (continues #809)

henningkayser opened the pull request:
This is an effort to finish PR #809 by @mikeferguson for integrating the [time-optimal trajectory parameterization](https://github.com/tobiaskunz/trajectories) method (published [here](http://www.golems.org/projects/traj.html)).

The branch is retargeted to melodic and I implemented a function for computing acceleration values (the original version doesn't do this).

Left TODOs:
- [x] process pending review by @davetcoleman 
- [x] extend API to support `tolerance` and `resample` parameters
- [x] debug edge cases mentioned in the original comments section


 \- Sponsored by Acutronic and PickNik -  


### Time-optimal Trajectory Parameterization
![totg](https://user-images.githubusercontent.com/4572766/53447189-ef4bf400-3a14-11e9-8195-5a0b2b12bcac.png)
### Iterative Spline Parameterization
![isp](https://user-images.githubusercontent.com/4572766/53447242-05f24b00-3a15-11e9-8f35-2486f0aa8bb5.png)
### Iterative Parabolic Time Parameterization
![iptp](https://user-images.githubusercontent.com/4572766/53447264-0f7bb300-3a15-11e9-96ea-dae6410657c7.png)

davetcoleman commented:
Thanks @henningkayser !

>  extend API to support tolerance and resample parameters

Is this feature necessary for this first PR? I'd rather just get the very stale PR merged so it gets into moveit2, even if its not feature complete. We don't want huge PRs.

mikeferguson commented:
@henningkayser thanks for picking this up and getting it across the line!

mikeferguson commented:
I'd also make one recommendation to maintainers -- don't squash this before merging, so that somebody can still figure out the changes we made from the original codebase (in case somebody wants to later import patches or whatnot).

henningkayser commented:
@davetcoleman I addressed your requested changes from #809. However, I would suggest we don't separate the classes into different files. Since we don't have any other code depending on it, separate files would just make the package more complicated without any real upside.

I also added some improvements to the implementation:
* The timestamp of the first waypoint is set to 0 (instead of a certain dt before)
* The last waypoint is always sampled from the last point in the processed path, fixing occasional non-zero end velocities
* trajectories that only contain the same waypoints are handled (fixes issue with planning to the current state)
* computeTimeStamps is now a member function of a 'TimeOptimalTrajectoryGeneration' class that allows setting tolerance and sampling parameters in the constructor

@hartmanndennis can you confirm that the issues you described are fixed?

> I'd also make one recommendation to maintainers -- don't squash this before merging, so that somebody can still figure out the changes we made from the original codebase (in case somebody wants to later import patches or whatnot).
I agree with @mikeferguson but I should probably squash my commits before merging.

davetcoleman commented:
> However, I would suggest we don't separate the classes into different files.

No problem

davetcoleman commented on moveit_core/trajectory_processing/include/moveit/trajectory_processing/time_optimal_trajectory_generation.h:
```diff
@@ -0,0 +1,179 @@
+  const double time_step_;
+
+  mutable double cached_time_;
+  mutable std::list<TrajectoryStep>::const_iterator cached_trajectory_segment_;
```
why are these mutable? this should be avoided

henningkayser replied:
These are mutable to allow caching results to speed up the const function `getTrajectorySegment()`. It's probably better than removing this feature or making the getter function non-const.

davetcoleman replied:
what if you have two instances of this class in different parts of the code base and sharing this memory would be a bug? this is what member variables are for

davetcoleman replied:
GStyle Guide:

> The mutable keyword is allowed but is unsafe when used with threads, so thread safety should be carefully considered first.

henningkayser replied:
These are member variables of Trajectory and two instances of Trajectory never share any memory.
The iterator `cached_trajectory_segment_` is generated from the private member list `trajectory_` (which never changes after the constructor) and both mutables are only changed inside `getTrajectorySegment()`. I don't know any way this could cause trouble. Do you see a better alternative for removing the mutable specifier than making all 'getter()' functions non-const?

davetcoleman replied:
Ah I think I see what you're saying. I'd still like to get @rhaschke 's opinion on this.

rhaschke replied:
Cache variables are typical candidates for the mutable keyword. I fully agree with Henning, that these variables are save here as long as the Trajectory instance isn't accessed in parallel from multiple threads. But the chances for this are really low, aren't they?

henningkayser replied:
Correct, this doesn't happen at all since each Trajectory instance only lives inside `computeTimeStamps()` which doesn't create multiple threads.

davetcoleman replied:
+1

davetcoleman commented on moveit_core/trajectory_processing/include/moveit/trajectory_processing/time_optimal_trajectory_generation.h:
```diff
@@ -0,0 +1,179 @@
+  std::list<TrajectoryStep> trajectory_;
+  std::list<TrajectoryStep> end_trajectory_;  // non-empty only if the trajectory generation failed.
+
+  static constexpr double EPS = 0.000001;
```
move into cpp file, at top of file

davetcoleman commented on moveit_core/trajectory_processing/src/time_optimal_trajectory_generation.cpp:
```diff
@@ -0,0 +1,1018 @@
+
+namespace trajectory_processing
+{
+const std::string LOGNAME = "trajectory_processing.time_optimal_trajectory_generation";
```
```suggestion
const std::string LOGNAME = "time_optimal_trajectory_generation";
```

henningkayser replied:
This follows the pattern from the other parameterization classes

davetcoleman replied:
no problem

davetcoleman commented on moveit_core/trajectory_processing/src/time_optimal_trajectory_generation.cpp:
```diff
@@ -0,0 +1,1018 @@
+{
+  if (path.size() < 2)
+    return;
+  std::list<Eigen::VectorXd>::const_iterator config1 = path.begin();
```
```suggestion
  std::list<Eigen::VectorXd>::const_iterator path_iterator1 = path.begin();
```

davetcoleman commented on moveit_core/trajectory_processing/src/time_optimal_trajectory_generation.cpp:
```diff
@@ -0,0 +1,1018 @@
+  if (path.size() < 2)
+    return;
+  std::list<Eigen::VectorXd>::const_iterator config1 = path.begin();
+  std::list<Eigen::VectorXd>::const_iterator config2 = config1;
```
```suggestion
  std::list<Eigen::VectorXd>::const_iterator path_iterator2 = path_iterator1;
```

davetcoleman commented on moveit_core/trajectory_processing/src/time_optimal_trajectory_generation.cpp:
```diff
@@ -0,0 +1,1018 @@
+  std::list<Eigen::VectorXd>::const_iterator config1 = path.begin();
+  std::list<Eigen::VectorXd>::const_iterator config2 = config1;
+  ++config2;
+  std::list<Eigen::VectorXd>::const_iterator config3;
```
```suggestion
  std::list<Eigen::VectorXd>::const_iterator path_iterator3;
```

davetcoleman commented on moveit_ros/planning/planning_request_adapter_plugins/src/add_time_optimal_parameterization.cpp:
```diff
@@ -0,0 +1,81 @@
+{
+using namespace trajectory_processing;
+
+/** @brief This adapter uses the ... */
```
fill in desc

davetcoleman commented:
@henningkayser can you squash some of your commits together, but not the original author's commit per @mikeferguson ?

henningkayser commented:
> @henningkayser can you squash some of your commits together, but not the original author's commit per @mikeferguson ?

Done

hartmanndennis commented:
> 
> @hartmanndennis can you confirm that the issues you described are fixed?
> 

Both issues seem to be fixed. Great Work!

davetcoleman commented:
Whoo hoo! @henningkayser I think it would be worth announcing this new feature on Discourse, giving credit also to @mikeferguson and original author Tobias Kuntz. Could you do that, including a pointer to the new tutorial?

mikeferguson commented:
Thanks for getting this across the line (and creating plots to show just how much better this algorithm is).

hartmanndennis commented:
Will this be cherry-picked to kinetic?

rhaschke commented:
> Will this be cherry-picked to kinetic?

I would say no. Maintaining both branches is a lot of work, and a subset of maintainers is now focussing on MoveIt for ROS2 already. You can compile MoveIt Melodic on Kinetic, if you like to have those changes in Kinetic.

gavanderhoorn commented:
@hartmanndennis: please also note that as nice as it is, TOTG is not jerk-limited, so if you plan on using it with KUKA robots over RSI, I don't expect it -- but I would really like to -- solve all issues there.

davetcoleman commented:
+1 to compiling the ``master`` branch with your ROS Kinetic environment - its compatible!

poripsasw commented on moveit_core/trajectory_processing/src/time_optimal_trajectory_generation.cpp:
```diff
@@ -0,0 +1,1020 @@
+      trajectory.push_back(TrajectoryStep(path_pos, path_vel));
+      return true;
+    }
+    else if (path_vel < 0.0)
```
Hi, I know this is as already merged code but is this right? My project has heavy deacceleration and my path points generated from compute_cartesian (which works with iterative spline) does not work with totp. I get the error msg : "Error while integrating forward: Negative path velocity". Why cant the path velocity be negative (sorry haven't looked through the code enough). Please let me know if this seems like a valid query, I'll open a git issue then.
